network/p2p/peerstore/peerstore.go

```diff
@@ -47,6 +73,283 @@ func NewPeerStore(addrInfo []*peer.AddrInfo) (*PeerStore, error) {
+func (ps *PeerStore) AddPersistentPeers(dnsAddresses []string, networkName string, role network.PhoneBookEntryRoles) {
+
+	for _, addr := range dnsAddresses {
+		info, err := PeerInfoFromDomainPort(addr)
+		if err != nil {
+			return
+		}
+		data, _ := ps.Get(info.ID, "addressData")
+		if data != nil {
+			// we already have this.
+			// Make sure the persistence field is set to true
+			ad := data.(addressData)
+			ad.persistent = true
+			_ = ps.Put("peerID", "addressData", data)
+
+		} else {
+			// we don't have this item. add it.
+			ps.AddAddrs(info.ID, info.Addrs, libp2p.PermanentAddrTTL)
+			entry := makePhonebookEntryData(networkName, role, true)
+			_ = ps.Put(info.ID, "addressData", entry)
+		}
+	}
+}
+
+// Length returns the number of addrs in peerstore
+func (ps *PeerStore) Length() int {
+	return len(ps.PeersWithKeys())
+}
+
+// makePhonebookEntryData creates a new address entry for provided network name and role.
+func makePhonebookEntryData(networkName string, role network.PhoneBookEntryRoles, persistent bool) addressData {
+	pbData := addressData{
+		networkNames:          make(map[string]bool),
+		recentConnectionTimes: make([]time.Time, 0),
+		role:                  role,
+		persistent:            persistent,
+	}
+	pbData.networkNames[networkName] = true
+	return pbData
+}
+
+func (ps *PeerStore) deletePhonebookEntry(peerID peer.ID, networkName string) {
+	data, err := ps.Get(peerID, "addressData")
+	if err != nil {
+		return
+	}
+	ad := data.(addressData)
+	delete(ad.networkNames, networkName)
+	if 0 == len(ad.networkNames) {
+		ps.ClearAddrs(peerID)
+		_ = ps.Put(peerID, "addressData", nil)
+	}
+}
+
+// AppendTime adds the current time to recentConnectionTimes in
+// addressData of addr
+func (ps *PeerStore) appendTime(peerID peer.ID, t time.Time) {
+	data, _ := ps.Get(peerID, "addressData")
+	ad := data.(addressData)
+	ad.recentConnectionTimes = append(ad.recentConnectionTimes, t)
+	_ = ps.Put(peerID, "addressData", ad)
+}
+
+// PopEarliestTime removes the earliest time from recentConnectionTimes in
+// addressData for addr
+// It is expected to be later than ConnectionsRateLimitingWindow
+func (ps *PeerStore) popNElements(n int, peerID peer.ID) {
+	data, _ := ps.Get(peerID, "addressData")
+	ad := data.(addressData)
+	ad.recentConnectionTimes = ad.recentConnectionTimes[n:]
+	_ = ps.Put(peerID, "addressData", ad)
+}
+
+func (ps *PeerStore) filterRetryTime(t time.Time, role network.PhoneBookEntryRoles) []string {
+	o := make([]string, 0, len(ps.Peers()))
+	for _, peerID := range ps.Peers() {
+		data, _ := ps.Get(peerID, "addressData")
+		if data != nil {
+			ad := data.(addressData)
+			if t.After(ad.retryAfter) && role == ad.role {
+				o = append(o, peerID.String())
+			}
+		}
+	}
+	return o
+}
+
+func shuffleSelect(set []string, n int) []string {
+	if n >= len(set) || n == network.GetAllAddresses {
+		// return shuffled copy of everything
+		out := slices.Clone(set)
+		shuffleStrings(out)
+		return out
+	}
```

Comment:
I think you can eliminate this special case by doing something like:
```go
if n > len(set) || n == network.GetAllAddresses {
    n = len(set)
}
... skip lines 331-334, but same from 336 to 339
        indexSample[i] = rand.Intn(n-i) + i
... etc
```

Reply:
you're probably right. I copied this from the current phonebook implementation. I'm not sure if it was written this way for some perf reasons. I'll leave it unchanged.